src/disk_io_thread.cpp

```diff
@@ -1074,9 +1088,12 @@ namespace libtorrent
 			// don't evict write jobs if at least one other thread
 			// is flushing right now. Doing so could result in
 			// unnecessary flushing of the wrong pieces
-			if (evict > 0 && m_stats_counters[counters::num_writing_threads] == 0)
+			if (evict > 0)
 			{
-				try_flush_write_blocks(evict, completed_jobs, l);
+				mutex::scoped_lock f(m_flush_mutex);
+				while (m_stats_counters[counters::num_writing_threads] != 0)
```

Comment:
We can't just skip the flushing. The other thread could itself be in `try_flush_write_blocks` so if we bail out here we may leave dirty buffers with nothing to flush them. This causes a deadlock if the dirty buffers exceed `m_max_use` and all `peer_connection`s are blocked waiting to allocate more buffers.

Posting a deferred job would work. I think I like that better than what's here now.